library/nmstate_l3_interface.py

```diff
@@ -0,0 +1,236 @@
+    aggregate_spec['name'] = dict(required=True)
+
+    # remove default in aggregate spec, to handle common arguments
+    remove_default_spec(aggregate_spec)
+
+    argument_spec = dict(
+        aggregate=dict(type='list', elements='dict', options=aggregate_spec),
+        purge=dict(default=False, type='bool'),
+        # not in net_* specification
+        debug=dict(default=False, type='bool'),
+    )
+
+    argument_spec.update(element_spec)
+
+    required_one_of = [['name', 'aggregate']]
+    mutually_exclusive = [['name', 'aggregate']]
+
+    result = dict(
+        changed=False,
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_one_of=required_one_of,
+        mutually_exclusive=mutually_exclusive,
+        # not yet supported by nmstate
+        supports_check_mode=False
+    )
+
+    if module.params['aggregate']:
+        # FIXME impelement aggregate
+        module.fail_json(msg='Aggregate not yet supported', **result)
+
+    previous_state = netinfo.show()
+    interfaces = previous_state['interfaces']
+    name = module.params['name']
+
+    interface_state = get_interface_state(interfaces, name)
+
+    if module.params['state'] == 'present':
+        if not interface_state:
+            module.fail_json(msg='Interface "%s" not found' % (name,),
+                             **result)
+
+        interface_state = set_ipv4_addresses(interface_state,
+                                             module.params['ipv4'],
+                                             purge=module.params['purge'])
+
+    elif module.params['state'] == 'absent':
+        if interface_state:
+            if module.params['ipv4']:
+                # FIXME: Allow to remove single IP addresses
+                module.fail_json(
+                    msg='Removing individual addresses not supported',
+                    **result)
+            else:
+                ipconfig = interface_state.setdefault('ipv4', {})
+                ipconfig['enabled'] = False
+
+        # assume success when interface to configure is not present
+
+    interfaces = [interface_state]
+    new_partial_state = {'interfaces': interfaces}
+
+    if module.params.get('debug'):
+        debugfile, debugname = tempfile.mkstemp(
+            prefix='nmstate_l3_interface_debug-{}-'.format(int(time.time())))
+        debugfile = os.fdopen(debugfile, "w")
+        debugfile.write(json.dumps(new_partial_state, indent=4))
+        result['debugfile'] = debugname
+
+    if module.check_mode:
+        new_full_state = previous_state.update(new_partial_state)
+        result['state'] = new_full_state
+
+        # TODO: maybe compare only the state of the defined interfaces
+        if previous_state != new_full_state:
+            result['changed'] = True
+
+        return result
+    else:
+        netapplier.apply(new_partial_state)
+    current_state = netinfo.show()
+    if current_state != previous_state:
```

Comment:
This would only happen when something else changed the state after the module. I was wondering, maybe `netapplier.apply()` could return the state and an indication whether it changed the state, it is probably useful for other consumers of nmstate as well.

Reply:
Good idea. Can you open a ticket on this?